### PR TITLE
Import HPD Violations

### PR DESCRIPTION
**Pull request by rdlester**
`import_hpd_violations` imports the raw violations data that is available (and in the standard csv format) from HPD, and automatically handles incremental update; `import_hpd_open_violations` imports the current open violations dump. The former is added to the luigi_master import.

### CODE REVIEW

**Comment by rdlester** [comment]
Also ping @aepyornis

**Comment by rdlester** [comment]
Fixes #112

**Comment by rdlester** [comment]
First steps towards #71

File: data-imports/clean_utils.py [hunk 1]
```diff
@@ -62,6 +64,10 @@ def clean_boro(table, column, replacements):
 def add_boroid(table, column):
     """ Returns a SQL statement that adds a column for boro IDs (used in BBL) and
         populates via conversion from a cleaned boro column.
+
+    Args:
+        table
+        column - Name of the column containing cleaned two-letter boro code.
```
Comment (chrishenry) [inline data-imports/clean_utils.py:70]: 👍

File: data-imports/import_hpd_open_violations.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+def import_csv(args):
+    csv_dir = os.path.join(BASE_DIR, table_name)
+    mkdir_p(csv_dir)
+
+    zip_file = os.path.join(csv_dir, "hpd_violations.zip")
+    csv_file = os.path.join(csv_dir, "Violation20161031.txt")
+
+    if not os.path.isfile(csv_file) or args.BUST_DISK_CACHE:
+        log.info("DL-ing HPD Buildings")
+        download_file("http://www1.nyc.gov/assets/hpd/downloads/misc/AllOpenViolations20161101.zip", zip_file)
```
Comment (chrishenry) [inline data-imports/import_hpd_open_violations.py:98]: That's the weirdest thing; why would they name a zip file with one date, but have it contain a different one.

File: data-imports/import_hpd_violations.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
+import calendar
+import clean_utils
+import datetime
+import luigi
+import luigi_utils
+import os
+import tempfile
+import utils
+
+class DownloadHPDViolations():
```
Comment (chrishenry) [inline data-imports/import_hpd_violations.py]: Small nitpick, but we [should be using new-style classes](https://stackoverflow.com/questions/54867/what-is-the-difference-between-old-style-and-new-style-classes-in-python) the inherit from `object`

File: data-imports/import_hpd_violations.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            'NOVID': 'object',
+            'NOVDescription': 'object',
+            'NOVIssuedDate': 'object',
+            'CurrentStatus': 'object',
+            'CurrentStatusDate': 'object'
+            # 'BoroID': 'object',
+            # 'Apartment': 'object',
+            # 'Story': 'object',
+            # 'ApprovedDate': 'object',
+            # 'CurrentStatusID': 'int64',
```
Comment (chrishenry) [inline data-imports/import_hpd_violations.py:101]: Why are these commented? Some look important, like `BoroID`

**Review by chrishenry — APPROVED** [review]
Looks good. I like the new class structure for hpd_violations.

File: data-imports/import_hpd_open_violations.py (hunk 2 above)
Comment (rdlester) [reply data-imports/import_hpd_open_violations.py:98]: My guess: The zip is released on the first of the month, but the data only runs til the last day of the previous month.

File: data-imports/import_hpd_violations.py (hunk 3 above)
Comment (rdlester) [reply data-imports/import_hpd_violations.py]: Done.

File: data-imports/import_hpd_violations.py (hunk 4 above)
Comment (rdlester) [reply data-imports/import_hpd_violations.py:101]: See the comment on line 70 - they weren't included in the first v2 (txt / csv) updates from 2013, but are in the most recent updates. I didn't figure out when they started getting included. I guess I could try something like creating a method that returns different dtype_dicts for a provided date, but then we'd need to also separate database schema specification from the csv data description. Certainly doable, and maybe worth doing, given that I've been running into memory limits when trying to do stuff with this violations data.

**Comment by rdlester** [comment]
I added some more comments + some fixes to the cleaning function (my computer was running out of memory when trying to run the cleaning, but I also only have 5gb left on my computer).

File: data-imports/clean_utils.py [hunk 5]
```diff
@@ -75,8 +75,8 @@ def add_boroid(table, column):
     if not table_exists(lookup_table):
         sql = sql + """
         CREATE TABLE {table} (
-            boro_code varchar(2),
-            boro_id int,
+            boro_code char(2),
+            boro_id tinyint,
```
Comment (chrishenry) [inline data-imports/clean_utils.py:79]: Nice.

File: data-imports/import_hpd_violations.py (hunk 4 above)
Comment (chrishenry) [reply data-imports/import_hpd_violations.py:101]: Ah, got it.

Captured https://github.com/chrishenry/deltanyc-heatseek/issues/178

**Pull request merged**